kotlinx-coroutines-core/common/src/AbstractCoroutine.kt

```diff
@@ -102,15 +102,17 @@ public abstract class AbstractCoroutine<in T>(
             onCompleted(state as T)
     }
 
-    internal open val defaultResumeMode: Int get() = MODE_ATOMIC_DEFAULT
-
     /**
      * Completes execution of this with coroutine with the specified result.
      */
     public final override fun resumeWith(result: Result<T>) {
-        makeCompletingOnce(result.toState(), defaultResumeMode)
+        val state = makeCompletingOnce(result.toState())
+        if (state === COMPLETING_WAITING_CHILDREN) return
+        afterResume(state)
     }
 
+    protected open fun afterResume(state: Any?) = afterCompletion(state)
```

Comment:
The key point of this method is that it is overridden in `UndispatchedCoroutine`. This coroute performs a direct resume of its parent continuation in its `afterResume`, because it is called in the same context, but does dispatch in its `afterCompletion` method, because it is called in an unknown context.